src/runtime/storageNG/direct-store.ts

```diff
@@ -0,0 +1,241 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {CRDTModel, CRDTTypeRecord, CRDTChange, ChangeType, CRDTError} from '../crdt/crdt.js';
+import {Type} from '../type.js';
+import {Exists, Driver, DriverFactory} from './drivers/driver-factory.js';
+import {StorageKey} from './storage-key.js';
+import {ActiveStore, ProxyCallback, StorageMode, ProxyMessageType, ProxyMessage} from './store-interface.js';
+
+export enum DirectStoreState {Idle = 'Idle', AwaitingResponse = 'AwaitingResponse', AwaitingResponseDirty = 'AwaitingResponseDirty', AwaitingDriverModel = 'AwaitingDriverModel'}
+
+export class DirectStore<T extends CRDTTypeRecord> extends ActiveStore<T> {
+  localModel: CRDTModel<T>;
+  callbacks = new Map<number, ProxyCallback<T>>();
+  driver: Driver<T['data']>;
+  private nextCallbackID = 1;
+  private version = 0;
+  private pendingException: Error | null = null;
+  private pendingResolves: Function[] = [];
+  private pendingRejects: Function[] = [];
+  private pendingDriverModels: {model: T['data']; version: number;}[] = [];
+  private state: DirectStoreState = DirectStoreState.Idle;
+  /*
+   * This class should only ever be constructed via the static construct method
+   */
+  private constructor(storageKey: StorageKey, exists: Exists, type: Type, mode: StorageMode, modelConstructor: new () => CRDTModel<T>) {
```

Comment:
I think you can just delete this constructor, if it doesn't do anything but call its super constructor?

Reply:
This ensures that only DirectStore.construct can construct DirectStores, via the `private` keyword.

Reply:
Gotcha